PERF: Speed up Spearman calculation

dsaxton opened the pull request:
Saves off ranks for later reuse once they're calculated during the computation of `DataFrame.corr(method="spearman")` so as to speed up the calculation.

~~This PR ranks data outside the nested loop in `nancorr_spearman` so as to speed up the calculation~~ (closes #28139).  The performance benchmarks show a nice improvement:

```
asv continuous -f 1.1 upstream/master corr-perf -b stat_ops.Correlation

       before           after         ratio
     [5d9fd7e3]       [08d8e394]
     <master>         <corr-perf>
-      87.8±0.9ms       7.81±0.2ms     0.09  stat_ops.Correlation.time_corr('spearman', False)
-        88.0±1ms       7.82±0.1ms     0.09  stat_ops.Correlation.time_corr('spearman', True)
```

~~(One caveat here is that when two columns have different missing patterns we actually need to recompute ranks over the rows where both are non-missing, so in some situations this will actually perform slightly worse than the current implementation.)~~

Commit:
Rank outside loop

Commit:
Check for same missing pattern

Commit:
Add comment

f-vert commented:
does data analyst need a bachelors ?

On Mon, Aug 26, 2019 at 3:32 PM Daniel Saxton <notifications@github.com>
wrote:

> so in some situations this will actually perform slightly worse than the
> current implementation
>
> Is there a good heuristic that we can use to check this?
>
> The worst case would be when every column has a different missing pattern
> and the first step of precalculating is for naught, and in this case you do
> an extra k sorts. I think this is a good tradeoff since in the best case
> you do k instead of k**2 sorts.
>
> When I run that scenario on this branch I do get worse performance (4.74
> seconds vs. 4.49 seconds using a 5000 by 100 array), but it's actually
> almost within a margin of error of the current implementation.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/pandas-dev/pandas/pull/28151?email_source=notifications&email_token=AKHFDSCIQMRPYRIYLB6NOLLQGRKYLA5CNFSM4IPQM5WKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD5F4FCA#issuecomment-525058696>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AKHFDSB2X4JPHIYCT4KU3VDQGRKYLANCNFSM4IPQM5WA>
> .
>

Commit:
Avoid unnecessary ranking

Commit:
Rename same_miss_pat

WillAyd commented:
Can you add a benchmark that injects missing data into the calculation? I think it would help frame the discussion around performance a little better if that's something that needs to be accounted for

WillAyd commented on pandas/_libs/algos.pyx:
```diff
@@ -307,26 +308,44 @@ def nancorr_spearman(const float64_t[:, :] mat, Py_ssize_t minp=1):
     mask = np.isfinite(mat).view(np.uint8)
 
+    ranked_mat = np.empty((N, K), dtype=np.float64)
+    cached_ranks = set({})
```
IIUC the whole caching thing you have here is to avoid the memory overhead of ranking columns up front right? If so I *think* it's a premature optimization so would probably just prefer if everything was ranked and we didn't do caching.

dsaxton replied:
So this approach would accept a hit on memory by preallocating `ranked_mat` (even this is probably not necessary), but save time by not calculating ranks up front that we end up not reusing.  Instead we wait until we need a full column set of ranks for _some_ calculation before computing and saving them.

WillAyd replied:
How much of a time savings is that? In any case your PR should offer a significant speed up by not ranking in each loop so I think this *might* still be OK.

dsaxton replied:
It was actually fairly negligible based on some local testing I did (maybe 5% slowdown), but it's probably worthwhile to add some more cases to `stat_ops.py` (wide tables, possibly a worst case missing pattern for this change, etc.)

WillAyd replied:
Did you see any perf difference here? Again would rather simplify and not have this if there's no major perf impact

dsaxton replied:
I'm actually seeing essentially the same performance from the previous implementation (ranking everything up front) on the "wide with nans" benchmark, so maybe just revert to that approach since the code is a lot easier to understand?

jreback reviewed (changes requested):
do the existing benchmarks cover enough here? (e.g. enough matrix sizes)
pls add a perf note

Commit:
Add wide frame to benchmarks

Commit:
Add performance note

Commit:
Add peakmem benchmark

Commit:
Add nans to benchmark data

dsaxton commented:
Added a couple benchmarks, one with wide data (1000 by 500), wide data with random `nans`, and a memory benchmark on the wide data.  Here's the summary:

```
       before           after         ratio
     [498f3008]       [420b1d6a]
     <master>         <corr-perf>
-        84.1±2ms       7.86±0.5ms     0.09  stat_ops.Correlation.time_corr('spearman', False)
-      83.1±0.5ms       7.57±0.4ms     0.09  stat_ops.Correlation.time_corr('spearman', True)
-       23.3±0.1s       1.86±0.05s     0.08  stat_ops.Correlation.time_corr_wide('spearman', True)

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.
```

Couple things that didn't make it into the summary but are interesting:

`master`

(Not directly relevant to this pull request, but Kendall's tau fails on the wide data with `nans`.)

```
[ 75.00%] · For pandas commit 498f3008 <master> (round 2/2):
[ 75.00%] ·· Building for conda-py3.6-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt....
[ 75.00%] ·· Benchmarking conda-py3.6-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt
[ 78.57%] ··· stat_ops.Correlation.peakmem_corr_wide                                                                                                                     ok
[ 78.57%] ··· ========== ======= =======
              --          use_bottleneck
              ---------- ---------------
                method     True   False 
              ========== ======= =======
               spearman   82.2M   82.2M 
               kendall    98.7M   98.7M 
               pearson    81.8M   81.8M 
              ========== ======= =======

[ 92.86%] ··· stat_ops.Correlation.time_corr_wide_nans                                                                                                           2/6 failed
[ 92.86%] ··· ========== ============ ============
              --               use_bottleneck     
              ---------- -------------------------
                method       True        False    
              ========== ============ ============
               spearman    20.1±0s     20.0±0.02s 
               kendall      failed       failed   
               pearson    1.43±0.02s   1.55±0.06s 
              ========== ============ ============
```

`corr-perf`

Surprisingly the peak memory usage is not much worse than `master`.  If we add random null values to the data then the time performance is the same as `master`.

```
[ 50.00%] · For pandas commit 420b1d6a <corr-perf> (round 2/2):
[ 50.00%] ·· Benchmarking conda-py3.6-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt
[ 53.57%] ··· stat_ops.Correlation.peakmem_corr_wide                                                                                                                     ok
[ 53.57%] ··· ========== ======= =======
              --          use_bottleneck
              ---------- ---------------
                method     True   False 
              ========== ======= =======
               spearman    86M     86M  
               kendall    98.7M   98.7M 
               pearson    81.7M   81.6M 
              ========== ======= =======

[ 67.86%] ··· stat_ops.Correlation.time_corr_wide_nans                                                                                                           2/6 failed
[ 67.86%] ··· ========== ============ ============
              --               use_bottleneck     
              ---------- -------------------------
                method       True        False    
              ========== ============ ============
               spearman   20.1±0.1s     20.0±0s   
               kendall      failed       failed   
               pearson    1.42±0.02s   1.41±0.03s 
              ========== ============ ============
```

Commit:
Merge branch 'master' into corr-perf

WillAyd commented:
You might need to skip the Kendall method in benchmarks where failing to get CI to pass. can you open up a follow up issue to address?

dsaxton commented:
> You might need to skip the Kendall method in benchmarks where failing to get CI to pass. can you open up a follow up issue to address?

Is there an easy way to skip or xfail a benchmark in `asv`?

WillAyd commented:
Try raising a `NotImplementedError`

Commit:
Revert "Avoid unnecessary ranking"

This reverts commit 14e7f5166ef28dc71763e1a3b005f080e256e39c.

Commit:
Edit comment

Commit:
Raise for Kendall benchmark

Commit:
Try returning None

dsaxton commented:
> Try raising a `NotImplementedError`

I was seeing the benchmark still fail after doing this.  Feels like a hack, but how about returning `None` (or maybe making the data just small enough so they all pass)?

Commit:
Fix benchmark

dsaxton commented:
@WillAyd I lowered the dimensions of the wide benchmarks to 1000 by 300 and nothing fails now; also reverted to the simpler implementation since the performance was about the same.  Updated `asv` results:

`corr-perf`

```
[ 54.17%] ··· stat_ops.Correlation.time_corr                                                                                                                             ok
[ 54.17%] ··· ========== ============ ============
              --               use_bottleneck     
              ---------- -------------------------
                method       True        False    
              ========== ============ ============
               spearman   7.68±0.6ms   8.13±0.4ms 
               kendall     193±8ms      193±5ms   
               pearson    2.04±0.1ms   1.99±0.2ms 
              ========== ============ ============

[ 62.50%] ··· stat_ops.Correlation.time_corr_wide                                                                                                                        ok
[ 62.50%] ··· ========== =========== ===========
              --              use_bottleneck    
              ---------- -----------------------
                method       True       False   
              ========== =========== ===========
               spearman    666±30ms    674±20ms 
               kendall    20.0±0.1s   19.7±0.1s 
               pearson     429±30ms    456±20ms 
              ========== =========== ===========

[ 66.67%] ··· stat_ops.Correlation.time_corr_wide_nans                                                                                                                   ok
[ 66.67%] ··· ========== ============ ============
              --               use_bottleneck     
              ---------- -------------------------
                method       True        False    
              ========== ============ ============
               spearman   7.29±0.06s   7.72±0.07s 
               kendall    20.5±0.4s    20.2±0.7s  
               pearson     561±8ms      562±30ms  
              ========== ============ ============
```

`master`

```
[ 79.17%] ··· stat_ops.Correlation.time_corr                                                                                                                             ok
[ 79.17%] ··· ========== ============= ============
              --               use_bottleneck      
              ---------- --------------------------
                method        True        False    
              ========== ============= ============
               spearman     90.7±2ms     90.2±3ms  
               kendall     193±0.9ms     195±3ms   
               pearson    1.90±0.07ms   1.88±0.1ms 
              ========== ============= ============

[ 87.50%] ··· stat_ops.Correlation.time_corr_wide                                                                                                                        ok
[ 87.50%] ··· ========== ============ ============
              --               use_bottleneck     
              ---------- -------------------------
                method       True        False    
              ========== ============ ============
               spearman   8.43±0.08s   8.46±0.1s  
               kendall    19.3±0.03s   19.4±0.06s 
               pearson     438±20ms     445±20ms  
              ========== ============ ============

[ 91.67%] ··· stat_ops.Correlation.time_corr_wide_nans                                                                                                                   ok
[ 91.67%] ··· ========== ============ ===========
              --              use_bottleneck     
              ---------- ------------------------
                method       True        False   
              ========== ============ ===========
               spearman   7.23±0.01s   7.55±0.1s 
               kendall    19.2±0.08s   19.0±0.2s 
               pearson     505±20ms     532±20ms 
              ========== ============ ===========
```

jreback commented:
pls decrease the size even more

20s is too long

Commit:
Update benchmark data size

dsaxton commented:
> pls decrease the size even more
> 
> 20s is too long

Updated to 200 columns, longest benchmark takes ~8 seconds

jreback commented:
lgtm. over to you @WillAyd 

@dsaxton could even reduce the benchmark time further as the relative statistics are still telling. (and we have a ton of benchmarks; it already takes 1 hour + to run them all...)

WillAyd commented:
Thanks @dsaxton very nice PR. Would certainly love any follow ups or improvements you can offer!
